File: framework/languages/dotnet/samples/Bingo/Server/Play/Handlers/BingoRoomDirectory.cs
using Bingo.Server.Play.BingoRoomSpots;
using Bingo.Shared.Configuration;

namespace Bingo.Server.Play;

internal sealed class BingoRoomDirectory(IZLinkSpotManager spots)
{
    private readonly SemaphoreSlim _gate = new(1, 1);
    private string? _currentRoomId;
    private BingoRoomSettings? _currentRoomSettings;
    private int _reservedSeats;
    private int _roomSeq;

    public async ValueTask<string> AllocateAsync(
        string mode,
        CancellationToken cancellationToken)
    {
        var settings = BingoRoomSettings.Create(mode, _roomSeq + 1);

        await _gate.WaitAsync(cancellationToken);
        try
        {
            if (_currentRoomId is null
                || _currentRoomSettings is null
                || !string.Equals(_currentRoomSettings.Mode, settings.Mode, StringComparison.Ordinal)
                || _reservedSeats >= _currentRoomSettings.RequiredPlayers)
            {
                settings = settings with { RoomName = $"Bingo Room {++_roomSeq:000}" };
                using var settingsPart = settings.ToJson();
                var room = await spots.CreateAsync(SampleNames.RoomSpotType, [settingsPart], cancellationToken)
                    ;
                _currentRoomId = room.SpotRid.ToHex();
                _currentRoomSettings = settings;
                _reservedSeats = 0;
            }

            _reservedSeats++;
            return _currentRoomId;
        }
        finally
        {
            _gate.Release();
        }
    }
}

File: framework/languages/node/samples/Bingo.Ts/Server/Play/Handlers/bingo-room-directory.ts
const { createRoomSettings } = require('../BingoRoomSpots/bingo-room-models');
const { BingoRoomSpot } = require('../BingoRoomSpots/bingo-room-spot');

class BingoRoomDirectory {
  [key: string]: any;
  constructor(notifications) {
    this.notifications = notifications;
    this.currentRoomId = null;
    this.currentRoomSettings = null;
    this.reservedSeats = 0;
    this.roomSeq = 0;
    this.rooms = new Map();
  }

  async allocate(mode) {
    let settings = createRoomSettings(mode, this.roomSeq + 1);
    if (
      this.currentRoomId === null ||
      this.currentRoomSettings === null ||
      this.currentRoomSettings.mode !== settings.mode ||
      this.reservedSeats >= this.currentRoomSettings.requiredPlayers
    ) {
      this.roomSeq += 1;
      settings = createRoomSettings(mode, this.roomSeq);
      this.currentRoomId = `bingo-room-${String(this.roomSeq).padStart(3, '0')}`;
      this.currentRoomSettings = settings;
      this.reservedSeats = 0;
      this.rooms.set(this.currentRoomId, new BingoRoomSpot(this.currentRoomId, settings, this.notifications));
    }

    this.reservedSeats += 1;
    return {
      roomId: this.currentRoomId,
      room: this.rooms.get(this.currentRoomId)
    };
  }

  require(roomId) {
    const room = this.rooms.get(roomId);
    if (room === undefined) {
      throw new Error(`Bingo room ${roomId} does not exist.`);
    }
    return room;
  }
}

export { BingoRoomDirectory };
